target/linux/ath79/dts/ar7100_mikrotik_routerboard-4xx.dtsi

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-1.0-or-later OR MIT
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "ar7100.dtsi"
+
+/ {
+	aliases {
+		led-boot = &led_user;
+		led-failsafe = &led_user;
+		led-running = &led_user;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_user: user {
+			label = "mikrotik:yellow:user";
+			gpios = <&gpio 4 GPIO_ACTIVE_HIGH>;
+			default-state = "on";
+		};
+
+		led1 {
+			label = "mikrotik:green:led1";
+			gpios = <&cpld_gpio 0 GPIO_ACTIVE_LOW>;
+		};
+
+		led2 {
+			label = "mikrotik:green:led2";
+			gpios = <&cpld_gpio 2 GPIO_ACTIVE_LOW>;
+		};
+
+		led3 {
+			label = "mikrotik:green:led3";
+			gpios = <&cpld_gpio 1 GPIO_ACTIVE_LOW>;
+		};
+
+		led4 {
+			label = "mikrotik:green:led4";
+			gpios = <&cpld_gpio 3 GPIO_ACTIVE_LOW>;
+		};
+
+		led5 {
+			label = "mikrotik:green:led5";
+			gpios = <&cpld_gpio 8 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&gpio {
+	ngpios = <31>;
+	gpio-line-names =
+		"",     "", "", "", "LED", "RDY",  "",  "MDC",
+		"MDIO", "", "", "", "",    "",     "",  "",
+		"",     "", "", "", "",    "",     "",  "",
+		"",     "", "", "", "",    "",     "",  "";
+};
+
+&pcie0 {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	compatible = "mikrotik,rb4xx-spi";
+
+	flash@0 {
+		compatible = "pm25lv512", "jedec,spi-nor";
+
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
+		partitions {
+			compatible = "mikrotik,routerboot-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "routerboot1";
+				reg = <0x0 0x0>;
+				read-only;
+			};
+
+			hard_config {
+				label = "hard_config";
+				read-only;
+			};
+
+			bios {
+				label = "bios";
+				size = <0x1000>;
+				read-only;
+			};
+
+			routerboot2 {
+				label = "routerboot2";
+				read-only;
+			};
+
+			soft_config {
+				label = "soft_config";
+			};
+		};
+	};
+
+	cpld@1 {
+		compatible = "mikrotik,rb4xx-cpld";
+
+		reg = <1>;
+		spi-max-frequency = <25000000>;
+
+		cpld_gpio: gpio {
+			compatible = "mikrotik,rb4xx-gpio";
+
+			base = <32>;
+
+			gpio-controller;
+			#gpio-cells = <2>;
+			gpio-line-names =
+				"LED1", "LED2", "LED3", "LED4",
+				"FAN",  "ALE",  "CLE",  "nCE",
+				"LED5";
+		};
+
+		nand {
+			compatible = "mikrotik,rb4xx-nand";
+
+			gpios = <&cpld_gpio 5 GPIO_ACTIVE_HIGH>, // ALE
+				<&cpld_gpio 6 GPIO_ACTIVE_HIGH>, // CLE
+				<&cpld_gpio 7 GPIO_ACTIVE_HIGH>, // nCE
+				<&gpio      5 GPIO_ACTIVE_HIGH>; // RDY
+
+			partitions {
+				compatible = "fixed-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
+					label = "booter";
+					reg = <0x0000000 0x0040000>;
+					read-only;
+				};
+
+				partition@40000 {
+					label = "kernel";
+					reg = <0x0040000 0x03c0000>;
```

Comment:
Quick data point on this: after discussing on IRC with rmilecki and others, it turns out that it is not technically feasible to have a dynamic kernel partition (it's not possible to move the head of the ubi partition as needed).
So even if we implement a top level "firmware" partition to hide the split, the kernel partition size would still be hardcoded for all practical purposes.

Given the above, and to try and remain as future-proof as possible with the kernel size inflation, and assuming the routerboot bootloader doesnt itself hardcode the kernel partition size (it doesn't seem to on NOR devices, where the kernel can live practically anywhere), and given that we have plenty of NAND space on these devices, I would make that partition significantly larger. Maybe `0x800000` or even `0x1000000` ? That way we could even use plain elf kernels and leave lzma-loader out of the equation (sparing us some memory overwrite headaches)? @xback @adrianschmutzler opinions?

Reply:
Go for it. This device/these devices give us so much to worry about, I'm open to anything that makes our lives a little easier (now or in the future).